README.md
# RTMP/RTSP/HLS Bridge for Wyze Cam

[![Docker](https://github.com/mrlt8/docker-wyze-bridge/actions/workflows/docker-image.yml/badge.svg)](https://github.com/mrlt8/docker-wyze-bridge/actions/workflows/docker-image.yml)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/mrlt8/docker-wyze-bridge?logo=github)](https://github.com/mrlt8/docker-wyze-bridge/releases/latest)
[![Docker Image Size (latest semver)](https://img.shields.io/docker/image-size/mrlt8/wyze-bridge?sort=semver&logo=docker&logoColor=white)](https://hub.docker.com/r/mrlt8/wyze-bridge)
[![Docker Pulls](https://img.shields.io/docker/pulls/mrlt8/wyze-bridge?logo=docker&logoColor=white)](https://hub.docker.com/r/mrlt8/wyze-bridge)

Docker container to expose a local RTMP, RTSP, and HLS stream for ALL your Wyze cameras including the outdoor and doorbell cams. No third-party or special firmware required.

It just works!

Based on [@noelhibbard's script](https://gist.github.com/noelhibbard/03703f551298c6460f2fd0bfdbc328bd#file-readme-md) with [kroo/wyzecam](https://github.com/kroo/wyzecam) and [aler9/rtsp-simple-server](https://github.com/aler9/rtsp-simple-server).

Please consider starring or [sponsoring](https://ko-fi.com/mrlt8) this project if you found it useful.

## Quick Start

Install [docker](https://docs.docker.com/get-docker/) and use your Wyze credentials to run:

```bash
docker run \
  -e WYZE_EMAIL=you@email.com \
  -e WYZE_PASSWORD=yourpassw0rd \
  -p 8888:8888 mrlt8/wyze-bridge:latest
```

You can view your stream by visiting: `http://localhost:8888/cam-nickname` where localhost is the hostname or ip of the machine running the bridge followed by the cam nickname in lowercase with `-` in place of spaces.

See [basic usage](#basic-usage) for additional information.

## Changes in v1.4.4

- **CHANGED**: 🚧 MQTT now reports camera `state` as "online", "offline", "disconnected", or the connection error.
- **CHANGED**: 🚧 MQTT now displays camera `net_mode`, `wifi`, and `audio`.

- **UPDATED**: ⬆️ rtsp-simple-server > [v0.18.3](https://github.com/aler9/rtsp-simple-server/releases/tag/v0.18.3)

## Changes in v1.4.2/3

- **FIXED**: 🔧 Bug in v1.4.2 if MQTT was enabled in home assistant. (#375) Thanks @JochenKlenk!

- **NEW**: ✨ ENV: `OFFLINE_IFTTT={event}:{key}` - Send a webhook trigger to IFTTT when the camera goes offline (-90).

- **CHANGED**: 🚧 MQTT now reports camera `state` as "connected", "disconnected", "offline", or the connection error. (#359)

- **FIXED**: 🔧 Use case-sensitive keys for livestream. (#371) Thanks @radnor!
- **FIXED**: 🔧 Stream would not come back when audio was enabled. (#347) Thanks @compeek!

[View previous changes](https://github.com/mrlt8/docker-wyze-bridge/releases)

## Features

- Access to video and audio for all Wyze-supported cameras via RTSP/RTMP/HLS.
- Access to HD *or* SD stream with configurable bitrate.
- Local and remote access to any of the cams on your account.
- Runs on almost any x64 or armv7/arm64 based system like a Raspberry Pi that supports docker.
- Support for Wyze 2FA.
- Ability to rotate video for Wyze Doorbell.
- Ability to record streams locally.
- Ability to take snapshots on an interval.
- Ability to livestream directly from the bridge.
- Ability to send a IFTTT webhook when a camera is offline (-90).

## Supported Cameras

![Wyze Cam V2](https://img.shields.io/badge/wyze_v2-yes-success.svg)
![Wyze Cam V3](https://img.shields.io/badge/wyze_v3-yes-success.svg)
![Wyze Cam Floodlight](https://img.shields.io/badge/wyze_floodlight-yes-success.svg)
![Wyze Cam Pan](https://img.shields.io/badge/wyze_pan-yes-success.svg)
![Wyze Cam Pan V2](https://img.shields.io/badge/wyze_pan_v2-yes-success.svg)
![Wyze Cam Outdoor](https://img.shields.io/badge/wyze_outdoor-yes-success.svg)
![Wyze Cam Outdoor V2](https://img.shields.io/badge/wyze_outdoor_v2-yes-success.svg)
![Wyze Cam Doorbell](https://img.shields.io/badge/wyze_doorbell-yes-success.svg)

![Wyze Cam v1](https://img.shields.io/badge/wyze_v1-no-inactive.svg)
![Wyze Cam Doorbell Pro](https://img.shields.io/badge/wyze_doorbell_pro-no-inactive.svg)

| Camera                | Model          | Supported |
| --------------------- | -------------- | --------- |
| Wyze Cam v1           | WYZEC1         | ⚠️         |
| Wyze Cam V2           | WYZEC1-JZ      | ✅         |
| Wyze Cam V3           | WYZE_CAKP2JFUS | ✅         |
| Wyze Cam Floodlight   | WYZE_CAKP2JFUS | ✅         |
| Wyze Cam Pan          | WYZECP1_JEF    | ✅         |
| Wyze Cam Pan v2       | HL_PAN2        | ✅         |
| Wyze Cam Outdoor      | WVOD1          | ✅         |
| Wyze Cam Outdoor v2   | HL_WCO2        | ✅         |
| Wyze Cam Doorbell     | WYZEDB3        | ✅         |
| Wyze Cam Doorbell Pro | GW_BE1         | ❓         |

### Firmware Compatibility

The bridge should be compatible with official firmware from wyze.

## Compatibility

![Supports armv7 Architecture](https://img.shields.io/badge/armv7-yes-success.svg)
![Supports aarch64 Architecture](https://img.shields.io/badge/aarch64-yes-success.svg)
![Supports amd64 Architecture](https://img.shields.io/badge/amd64-yes-success.svg)

[![Home Assistant Add-on](https://img.shields.io/badge/home_assistant-add--on-blue.svg?logo=homeassistant&logoColor=white)](https://github.com/mrlt8/docker-wyze-bridge/wiki/Home-Assistant)
[![Homebridge](https://img.shields.io/badge/homebridge-camera--ffmpeg-blue.svg?logo=homebridge&logoColor=white)](https://sunoo.github.io/homebridge-camera-ffmpeg/configs/WyzeCam.html)
[![Portainer stack](https://img.shields.io/badge/portainer-stack-blue.svg?logo=portainer&logoColor=white)](https://github.com/mrlt8/docker-wyze-bridge/wiki/Portainer)
[![Unraid Community App](https://img.shields.io/badge/unraid-community--app-blue.svg?logo=unraid&logoColor=white)](https://github.com/mrlt8/docker-wyze-bridge/issues/236)

Should work on most x64 systems as well as on some arm-based systems like the Raspberry Pi.

The container can be run on its own, in [Portainer](https://github.com/mrlt8/docker-wyze-bridge/wiki/Portainer), [Unraid](https://github.com/mrlt8/docker-wyze-bridge/issues/236), as a [Home Assistant Add-on](https://github.com/mrlt8/docker-wyze-bridge/wiki/Home-Assistant), locally or remotely in the cloud.

[![Open your Home Assistant instance and show the add add-on repository dialog with a specific repository URL pre-filled.](https://my.home-assistant.io/badges/supervisor_add_addon_repository.svg)](https://my.home-assistant.io/redirect/supervisor_add_addon_repository/?repository_url=https%3A%2F%2Fgithub.com%2Fmrlt8%2Fdocker-wyze-bridge)

## Basic Usage

### docker-compose (recommended)

This is similar to the docker run command, but will save all your options in a yaml file.

1. Install [Docker Compose](https://docs.docker.com/compose/install/).
2. Use the [sample](https://raw.githubusercontent.com/mrlt8/docker-wyze-bridge/main/docker-compose.sample.yml) as a guide to create a `docker-compose.yml` file with your wyze credentials.
3. Run `docker-compose up`.

Once you're happy with your config you can use `docker-compose up -d` to run it in detached mode.

#### Updating your container

To update your container, `cd` into the directory where your `docker-compose.yml` is located and run:

```bash
docker-compose pull # Pull new image
docker-compose up -d # Restart container in detached mode
docker image prune # Remove old images
```

### 🏠 Home Assistant

Visit the [wiki page](https://github.com/mrlt8/docker-wyze-bridge/wiki/Home-Assistant) for additional information on Home Assistant.

### Additional Info

- [Two-Step Verification](#Multi-Factor-Authentication)
- [ARM/Raspberry Pi](https://github.com/mrlt8/docker-wyze-bridge/wiki/Raspberry-Pi-(armv7-and-arm64))
- [LAN mode](#LAN-Mode)
- [Portainer](https://github.com/mrlt8/docker-wyze-bridge/wiki/Portainer)
- [Unraid](https://github.com/mrlt8/docker-wyze-bridge/issues/236)
- [Home Assistant](https://github.com/mrlt8/docker-wyze-bridge/wiki/Home-Assistant)
- [Homebridge Camera FFmpeg](https://sunoo.github.io/homebridge-camera-ffmpeg/configs/WyzeCam.html)
- [HomeKit Secure Video](https://github.com/mrlt8/docker-wyze-bridge/wiki/HomeKit-Secure-Video)

#### Special Characters

If your email or password contains a `%` or `$` character, you may need to escape them with an extra character. e.g., `pa$$word` should be entered as `pa$$$$word`

#### Camera Stream URIs

By default, the bridge will create three streams for each of your cameras which can be accessed at the following URIs, where `camera-nickname` is the name of the camera set in the Wyze app and converted to lower case with hyphens in place of spaces. e.g. 'Front Door' would be `/front-door`

Replace localhost with the hostname or ip of the machine running the bridge:

- RTMP:

  ```text
  rtmp://localhost:1935/camera-nickname
  ```

- RTSP:

  ```text
  rtsp://localhost:8554/camera-nickname
  ```

- HLS:

  ```text
  http://localhost:8888/camera-nickname/stream.m3u8
  ```

- HLS can also be viewed in the browser using:

  ```text
  http://localhost:8888/camera-nickname
  ```

#### Multi-Factor Authentication

Two-factor authentication ("Two-Step Verification" in the wyze app) is supported and will automatically be detected, however additional steps are required to enter your verification code.

- Echo the verification code directly to `/tokens/mfa_token` by opening a second terminal window and using:

  ```bash
  docker exec -it wyze-bridge sh -c 'echo "123456" > /tokens/mfa_token'
  ```

- Mount `/tokens/` locally and add your verification code to a file named `mfa_token`:

  ```YAML
  volumes:
      - ./tokens:/tokens/
  ```

- Generate Time-based One-Time Passwords:

  You can have the bridge auto generate and enter a Time-based One-Time Password (TOTP) by adding the secret key to the file `/tokens/totp` on standard installs or `/config/wyze-bridge/totp` for Home Assistant installs. You will need to create the file if it doesn't exist and mount it if necessary.

- 🏠 Home Assistant:

  Add your code to the text file: `/config/wyze-bridge/mfa_token.txt`.

- Portainer:
  
  Use the console to echo your code to the container:

  ```bash
  echo "123456" > /tokens/mfa_token
  ```

## Advanced Options

**WYZE_EMAIL** and **WYZE_PASSWORD** are the only two required environment variables.

**The following envs are all optional.**

### Audio

Audio is disabled by default and must be enabled in the ENV.

#### Enable audio

- For all cameras:

  ```YAML
  environment:
      ..
      - ENABLE_AUDIO=True
  ```

- For a specific camera:

  where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens:

  ```yaml
    - ENABLE_AUDIO_CAM_NAME=True
  ```

#### Audio output codec

By default, the bridge will attempt to copy the audio from the camera without re-encoding *unless* the camera is using a codec that isn't supported by RTSP, in which case the audio will be converted to AAC.

The `AUDIO_CODEC` ENV can be used should you need to re-encode the audio to another format for compatibility:

```yaml
  - AUDIO_CODEC=AAC
```

#### Audio filtering

Custom ffmpeg audio filters can be set with `AUDIO_FILTER`, but please note that audio filters will only be applied if re-encoding to another codec.

```yaml
  - AUDIO_FILTER=highpass=f=300,lowpass=f=2500,volume=volume=2
```

### Filtering

The default option will automatically create a stream for all the cameras on your account, but you can use the following environment options in your `docker-compose.yml` to filter the cameras.

All options are case-insensitivE, and take single or comma separated values.

#### Examples

- Whitelist by Camera Name (set in the wyze app):

  ```yaml
  environment:
      ..
      - FILTER_NAMES=Front Door, Driveway, porch cam
  ```

- Whitelist by Camera MAC Address:

  ```yaml
  - FILTER_MACS=00:aA:22:33:44:55, Aa22334455bB
  ```

- Whitelist by Camera Model:

  ```yaml
  - FILTER_MODELS=WYZEC1-JZ
  ```

- Whitelist by Camera Model Name:

  ```yaml
  - FILTER_MODELS=V2, v3, Pan
  ```

- Blacklisting:

  You can reverse any of these whitelists into blacklists by setting `FILTER_BLOCK`.

  ```yaml
  environment:
      ..
      - FILTER_NAMES=Bedroom
      - FILTER_BLOCK=true
  ```

### Network Connection Modes

Like the wyze app, the tutk library has three different modes to connect to the camera and will attempt to stream directly from the camera when on the same LAN as the camera in "LAN mode". If the camera is not available locally, it will either attempt to stream directly from your network using "P2P Mode" or relay the stream via the wyze servers (AWS) in "relay mode".

LAN mode is more ideal as all streaming will be local and won't use additional bandwidth.

#### LAN Mode

By default, the bridge will attempt to connect via "LAN Mode", but will fallback to other methods if LAN mode fails.
You can restrict streaming to LAN only by setting the `NET_MODE=LAN` environment variable:

```yaml
environment:
    ..
    - NET_MODE=LAN
```

#### P2P Mode

`NET_MODE=P2P` is ideal when running the bridge remotely on a different network or on a VPS and will allow the bridge to stream directly from the camera over the internet while blocking "Relay Mode".

#### ANY Mode

`NET_MODE=ANY` is the equivalent to leaving `NET_MODE` unset and will allow the connection to fallback to P2P or relay mode.

#### `NET_MODE` for a specific camera

In the event that you need to allow the bridge to access a select number of cameras outside of your LAN, you can specify them by appending the camera name to `NET_MODE`, where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens:

```yaml
    ..
    - NET_MODE=LAN
    - NET_MODE_CAM_NAME=P2P
```

### Snapshot/Still Images

- `SNAPSHOT=API` Will run ONCE at startup and will grab a *high-quality* thumbnail from the wyze api and save it to `/img/cam-name.jpg` on docker installs or `/config/www/cam-name.jpg` in Home Assistant mode.

- `SNAPSHOT=RTSP` Will run every 180 seconds (configurable) and wll grab a new frame from the RTSP stream every iteration and save it to `/img/cam-name.jpg` on standard docker installs or `/config/www/cam-name.jpg` in Home Assistant mode. Can specify a custom interval with `SNAPSHOT=RTSP(INT)` e.g. `SNAPSHOT=RTSP30` to run every 30 seconds

- `IMG_DIR=/img/` Specify the directory where the snapshots will be saved *within the container*. Use volumes in docker to map to an external directory.

- `IMG_TYPE` Specify the file type of the image, e.g. `IMG_TYPE=png`. Will default to jpg.

### Stream Recording

The bridge can be configured to record all or select camera streams to the container which can be mapped to a local directory.

#### Enable recording

```yaml
environment:
...
  - TZ=America/New_York
  - RECORD_ALL=True
volumes:
  - /local/path/:/record/
```

Or to specify select cameras, where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens:

```yaml
  - RECORD_CAM_NAME=True
  - RECORD_OTHER_CAM=True
```

See the [Stream Recording wiki page](https://github.com/mrlt8/docker-wyze-bridge/wiki/Stream-Recording#recording-configuration) page for additional options.

### Offline camera IFTTT webhook (BETA)

This option can send a trigger to [IFTTT's webhooks integration](https://ifttt.com/maker_webhooks) when the camera is detected as being offline (-90).

Note:This is still experimental as the connection will sometimes just timeout a number of times before going "offline".

```yaml
  - OFFLINE_IFTTT=MyEventName:my_IFTTT_Webhooks_Key
```

The bridge should then trigger your event with the following values:

```json
{
  "value1" : "cam-name", // camera name
  "value2" : -90, // error no
  "value3" : "IOTC_ER_DEVICE_OFFLINE" // error name
}
```

### Livestream

Basic livestream support is available for YouTube and Facebook, but you can also specify any custom rtmp server for other services like Twitch.

To use this feature, set a new env in your docker-compose.yml with the service (`YOUTUBE_` or `FACEBOOK_`) prefix followed by the camera name in UPPERCASE with `_` in place of spaces and hyphens, and set your stream key as the value. Custom rtmp servers can be specified using the `LIVESTREAM_` prefix:

```yaml
  - YOUTUBE_FRONT_DOOR=MY-STREAM-KEY
  - FACEBOOK_OTHER_CAM=MY-STREAM-KEY
  # twitch example:
  - LIVESTREAM_CAM_NAME=rtmp://jfk.contribute.live-video.net/app/MY-STREAM-KEY
```

### MQTT (beta)

Some basic MQTT support is now available in v0.7.0.

MQTT auth and discovery should be automatic in Home Assistant mode - can be disabled by setting `MQTT_HOST` to False.

| ENV Name    | Description                                   | Example             |
| ----------- | --------------------------------------------- | ------------------- |
| MQTT_HOST   | IP/Hostname AND Port of the MQTT broker       | core-mosquitto:1883 |
| MQTT_AUTH   | Username AND password; leave blank if none    | user:pass           |
| MQTT_TOPIC  | Optional - Specify topic prefix               | myhome              |
| MQTT_DTOPIC | Optional - Discovery topic for home assistant | homeassistant       |

### Bitrate and Resolution

Bitrate and resolution of the stream from the wyze camera can be adjusted with:

#### Set quality for all cameras

```yaml
environment:
    - QUALITY=HD120
```

#### Set quality for single camera

where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens:

```yaml
environment:
    - QUALITY_CAM_NAME=HD120
```

Additional info:

- Resolution can be set to `SD` (360p in the app) or `HD` - 640x360/1920x1080 for cams or 480x640/1296x1728 for doorbells.
- Bitrate can be set from 30 to 255. Some bitrates may not work with certain resolutions.
- Adjusting the bitrate and resolution in the bridge will also change the stream in the wyze app and vice versa.
- App equivalents would be:
  - 360p - SD30
  - SD - HD60 (HD120 for doorbell)
  - HD - HD120 (HD180 for doorbell)

### Custom FFmpeg Commands

You can pass a custom [command](https://ffmpeg.org/ffmpeg.html) to FFmpeg by using `FFMPEG_CMD` in your docker-compose.yml:

#### For all cameras

```YAML
environment:
    ..
    - FFMPEG_CMD=-f h264 -i - -vcodec copy -f flv rtmp://rtsp-server:1935/{cam_name}
```

#### For a specific camera

where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens:

```yaml
- FFMPEG_CMD_CAM_NAME=ffmpeg -f h264 -i - -vcodec copy -f flv rtmp://rtsp-server:1935/{cam_name}
```

Additional info:

- The `ffmpeg` command is implied and is optional.
- The camera name is available as a variable `{cam_name}` for lowercase and `{CAM_NAME}` for uppercase.
- The audio pipe and format are available as the variable `{audio_in}`.

### Custom FFmpeg Flags

Custom ffmpeg flags can easily be tested with:

```YAML
environment:
    ..
    - FFMPEG_FLAGS=-fflags +flush_packets+genpts+discardcorrupt+nobuffer
```

or where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens:

```yaml
- FFMPEG_FLAGS_CAM_NAME=-flags low_delay
```

### rtsp-simple-server

[rtsp-simple-server](https://github.com/aler9/rtsp-simple-server/blob/main/rtsp-simple-server.yml) options can be customized as an environment variable in your docker-compose.yml by prefixing `RTSP_` to the UPPERCASE parameter.

e.g. use `- RTSP_RTSPADDRESS=:8555` to overwrite the default `rtspAddress`.

or `- RTSP_PATHS_ALL_READUSER=123` to customize a path specific option like `paths: all: readuser:123`

For camera specific options with spaces in the name of the camera (e.g. `Front Door`), be sure to replace the spaces with the `URI_SEPARATOR` which defaults to `-`.  So `Front Door` would be represented as `FRONT-DOOR`, and `paths: Front Door: runOnReady: ffmpeg...` could be set in your docker-compose as:

```yaml
environment:
  ...
  - RTSP_PATHS_FRONT-DOOR_RUNONREADY=ffmpeg...
```

### Debugging options

environment options:

- `FRESH_DATA` (bool) Remove local cache and pull new data from wyze servers.

- `URI_SEPARATOR` (-|_|#) Customize the separator used to replace spaces in the URI; available values are `-`, `_`, or use `#` to remove spaces.

- `CONNECT_TIMEOUT` (int) Adjust the number of seconds to wait before timing out when connecting to camera. Default: `15`

- `KEEP_BAD_FRAMES` (bool) Keep frames that may be missing a keyframe or preceding frames.

- `IGNORE_OFFLINE` (bool) Ignore offline cameras until container restarts.

- `OFFLINE_TIME` (int) Customize the sleep time when a camera is offline. Default: `10`

- `DEBUG_FRAMES` (bool) Show all lost/incomplete frames.

- `DEBUG_LEVEL` (debug|info|warning|error) Adjust the level of upstream logging.

- `RTSP_READTIMEOUT` (str) Adjust the max number of seconds of missing frames allowed before a stream is restarted. Be sure to include the s after the number. Default: `20s`

- `RTSP_LOGLEVEL` (debug|info|warn) Adjust the verbosity of rtsp-simple-server; available values are "warn", "info", "debug".

- `SKIP_RTSP_LOG` (bool) Prevent "read" spam in the logs when using RTSP based snapshots. Works by only logginng clients that stay connected for longer than 3s.

- `DEBUG_FFMPEG` (bool) Enable additional logging from FFmpeg.

- `FORCE_FPS_CAM_NAME` (int) Force a specific camera to use a different FPS, where `CAM_NAME` is the camera name in UPPERCASE and `_` in place of spaces and hyphens.

- `FPS_FIX` (bool) Set camera parameter to match the actual FPS being sent by the camera. Potential fix slow/fast SD card and cloud recordings.

- `WEBRTC` (bool) Display WebRTC credentials for cameras.


app/CHANGELOG.md
## Changes in v1.4.4

- **CHANGED**: 🚧 MQTT now reports camera `state` as "online", "offline", "disconnected", or the connection error.
- **CHANGED**: 🚧 MQTT now displays camera `net_mode`, `wifi`, and `audio`.

- **UPDATED**: ⬆️ rtsp-simple-server > [v0.18.3](https://github.com/aler9/rtsp-simple-server/releases/tag/v0.18.3)

## Changes in v1.4.2/3

- **FIXED**: 🔧 Bug in v1.4.2 if MQTT was enabled in home assistant. (#375) Thanks @JochenKlenk!

- **NEW**: ✨ ENV: `OFFLINE_IFTTT={event}:{key}` - Send a webhook trigger to IFTTT when the camera goes offline (-90).

- **CHANGED**: 🚧 MQTT now reports camera `state` as "connected", "disconnected", "offline", or the connection error. (#359)

- **FIXED**: 🔧 Use case-sensitive keys for livestream. (#371) Thanks @radnor!
- **FIXED**: 🔧 Stream would not come back when audio was enabled. (#347) Thanks @compeek!

## Changes in v1.4.0/1

- **NEW**: 🔊 Audio is now available. [Details](https://github.com/mrlt8/docker-wyze-bridge#audio)

- **UPDATED**: ⬆️ rtsp-simple-server > [v0.18.2](https://github.com/aler9/rtsp-simple-server/releases/tag/v0.18.2)

- **FIXED**: 🔧 Doorbell rotation. (#362) Thanks @krystiancharubin!

## Changes in v1.3.8

Audio is also coming soon. Please check out the audio branch to report any issues.

### 🚧 Changed

- Fixed a bug where the doorbell would fall behind and drift out of sync. Thanks @krystiancharubin!

## Changes in v1.3.7

### ✨ NEW

- Support for Wyze Cam Outdoor v2! (#354) Thanks @urubos!

### 🚧 Changed

- Fixed bug where the add-on would not start in Home Assistant if hostname was not set. (#355) Thanks @cbrightly!
- Fixed bug where rtsp-simple-server would refuse connections if the camera name contained a special character. (#356) Thanks @JochenKlenk!
- Set default doorbell bitrate to 180.

### 🐛 Bugs

There is a known bug/issue with certain doorbells that drift out of sync due to the day/night fps change (#340).

## Changes in v1.3.6

### 🚧 Changed

- Fixed bug in Home Assistant config that was causing the add-onn not to load. (#351) Thanks @jdeath, @JochenKlenk!
- Fixed bug in ffmpeg command to use protocol specified in `RTSP_PROTOCOLS`. (#347) Thanks @AdiAbuAli!

## Changes in v1.3.4/v1.3.5

### ✨ NEW

- ENV option: `IMG_TYPE` - Specify the snapshot image file type, e.g. `IMG_TYPE=png`
- ENV option: `SKIP_RTSP_LOG` - Prevent "read" spam in logs when using RTSP based snapshots.

### 🚧 Changed

- Fixed bug in v1.3.4 that could cause the CPU to spike.
- Fixed bug in `FILTER_MODELS` ENV that wouldn't match certain cameras (#346). Thanks @ragenhe!
- Fixed bug that could cause the stream to block when changing resolution/bitrate midstream (#340).
- Update rtsp-simple-server to v0.18.1.
- Improved speed of RTSP based snapshots!
- Force keyframes every two seconds on doorbell rotation.
- Limit doorbell bitrate to 3,000kb/s.
- MQTT related code refactoring and cleanup of unused topics.
- API: Wyze app version number bump to 2.30.0.

## Changes in v1.3.3

### ✨ NEW

- Livestreaming option now available. [Details](https://github.com/mrlt8/docker-wyze-bridge#livestream)

### 🚧 Changed

- Update FFmpeg to 5.0.
- Update rtsp-simple-server to v0.18.0.
- Tweaked doorbell rotation command for performance. #330
- HA: make `SNAPSHOT` optional and add `RTSP5`. #336
- Tweaked FFmpeg commands to use [tee muxer](https://ffmpeg.org/ffmpeg-formats.html#tee).
- API: iOS version bump to 15.4.1.
- API: Wyze app version number bump to 2.29.2.

## Changes in v1.3.2

⚠️ Potentially breaking for custom FFMPEG commands.

- Fixed custom ffmpeg ENV for camera names with spaces. #332
- Camera name variable for custom ffmpeg commands with `{cam_name}` for lowercase and `{CAM_NAME}` for uppercase. #334
- Camera name variable for `RECORD_FILE_NAME` and `RECORD_PATH` with `{cam_name}` for lowercase and `{CAM_NAME}` for uppercase.
- Changed default `RECORD_PATH` to `/record/{CAM_NAME}` and to `/media/wyze/{CAM_NAME}` for Home Assistant.
  
## Changes in v1.3.1

### 🚧 Changed

- Adjusted sleep time between frames that could cause the stream to fall behind. (#330) Thanks @bbobrian, @dreondre, and everyone who helped with reporting and testing!
- Additional FFMPEG commands to help reduce lag.
- Fixed spaces in ENV/YAML so that they use `_` instead of `-`. Thanks @ronald-mendoza!
- Updated typos in README. Thanks @ronald-mendoza! (#332)
- Wyze app version number bump (2.29.1).


## Changes in v1.3.0

### ✨ NEW

- Recording directly in the bridge is now here! [Details](#recording-streams-beta).
  
  🏠 Default settings will save recordings to `/media/wyze/` in Home Assistant mode.

### 🚧 Changed

- Adjusted connection timeout #306 #319.
- Check bitrate every 500 frames to detect any external changes #320.
- Correct mismatched FPS camera parameter with ENV: `FPS_FIX`.
- Add sleep between frames to lower CPU usage.
- Fixed import error #324.
- IOS and wyze app version number bump.


app/rtsp_event.py
import sys

try:
    import os
    import signal
    import time
    from datetime import datetime as dt

except ImportError:
    sys.exit(1)


class RtspEvent:
    """Handle events from rtsp-simple-server."""

    def __init__(self) -> None:
        """Run the script until it receives a termination signal."""
        for sig in ("SIGQUIT", "SIGTERM", "SIGINT"):
            signal.signal(getattr(signal, sig), lambda n, f: self.clean_up())
        self.uri: str
        self.type: str
        self.state: int = 0
        self.__dict__.update(
            dict(zip(["uri", "type", "mac", "model", "firmware"], sys.argv[1:]))
        )
        self.mqtt_connect()

    def write_log(self, txt: str) -> None:
        """Format and print logging messages to stdout."""
        date = dt.now().strftime("%Y/%m/%d %X")
        print(date, f"[RTSP][{self.uri.upper()}] {txt}")

    def pub_start(self) -> None:
        """Handle a 'READY' event when publishing a stream to rtsp-simple-server."""
        self.write_log(f"✅ '/{self.uri}' stream is UP! (3/3)")
        self.send_mqtt("image", None)

        env_snap = os.getenv("SNAPSHOT", "NA").ljust(5, "0").upper()
        img_file = (
            os.getenv("IMG_PATH", "/img/")
            + self.uri
            + "."
            + env_bool("IMG_TYPE", "jpg")
        )
        if rtsp_snap := (env_snap[:4] == "RTSP"):
            try:
                from subprocess import Popen, TimeoutExpired
            except ImportError:
                rtsp_snap = False
            rtsp_addr = "127.0.0.1:8554"
            if env_bool(f"RTSP_PATHS_{self.uri.upper()}_READUSER"):
                rtsp_addr = (
                    env_bool(f"RTSP_PATHS_{self.uri.upper()}_READUSER")
                    + ":"
                    + env_bool(f"RTSP_PATHS_{self.uri.upper()}_READPASS")
                    + f"@{rtsp_addr}"
                )
            ffmpeg_cmd = (
                ["ffmpeg", "-loglevel", "error", "-threads", "1"]
                + ["-analyzeduration", "50", "-probesize", "50"]
                + ["-rtsp_transport", "tcp", "-i", f"rtsp://{rtsp_addr}/{self.uri}"]
                + ["-f", "image2", "-frames:v", "1", "-y", img_file]
            )
        while True:
            self.send_mqtt("state", "online")
            if rtsp_snap:
                ffmpeg_sub = Popen(ffmpeg_cmd)
                try:
                    ffmpeg_sub.wait(15)
                except TimeoutExpired:
                    ffmpeg_sub.kill()
                    self.write_log("snapshot timed out")
                    continue
            if os.path.exists(img_file) and os.path.getsize(img_file) > 1:
                with open(img_file, "rb") as img:
                    self.send_mqtt("image", img.read())
            time.sleep(int(env_snap[4:] if env_snap[4:].isdigit() else 0) or 180)

    def read_start(self) -> None:
        """Handle 'READ' events when a client starts consuming a stream fromrtsp-simple-server."""
        if env_bool("SKIP_RTSP_LOG") and (os.getenv("SNAPSHOT", "NONE")[:4] == "RTSP"):
            time.sleep(3)
        self.write_log("📖 New client reading ")
        self.send_mqtt(f"clients/{os.getpid()}", "reading")
        self.state = 1
        signal.pause()

    def mqtt_connect(self) -> None:
        """Connect to an MQTT if the env option is enabled."""
        self.mqtt_connected = False
        if not env_bool("MQTT_HOST"):
            return
        try:
            import paho.mqtt.client as mqtt
        except ImportError:
            return
        self.base = f"wyzebridge/{self.uri}/"
        if env_bool("MQTT_TOPIC"):
            self.base = f'{env_bool("MQTT_TOPIC")}/{self.base}'
        host = os.getenv("MQTT_HOST").split(":")
        self.mqtt = mqtt.Client()
        if env_bool("MQTT_AUTH"):
            auth = os.getenv("MQTT_AUTH").split(":")
            self.mqtt.username_pw_set(auth[0], auth[1] if len(auth) > 1 else None)
        if self.type == "READY":
            self.mqtt.will_set(self.base + "state", "disconnected", 0, True)
        if self.type == "READ":
            self.mqtt.will_set(self.base + f"clients/{os.getpid()}", None, 0, True)
        try:
            self.mqtt.connect(host[0], int(host[1] if len(host) > 1 else 1883), 60)
            self.mqtt.loop_start()
            self.mqtt_connected = True
        except Exception as ex:
            self.write_log(f"[MQTT] {ex}")

    def send_mqtt(self, topic: str, message: str) -> None:
        """Publish a message to the MQTT server."""
        if self.mqtt_connected:
            if message:
                self.mqtt.publish(self.base + topic, message)
            else:
                self.mqtt.publish(self.base + topic, None, 0, True)

    def clean_up(self) -> None:
        """Update the log and MQTT status when a termination signal is received."""
        if self.type == "READY":
            self.write_log(f"❌ '/{self.uri}' stream is down")
            self.send_mqtt("state", "disconnected")
            self.send_mqtt("attributes", None)
        elif self.type == "READ" and self.state:
            self.write_log("📕 Client stopped reading")
            self.send_mqtt(f"clients/{os.getpid()}", None)
        if self.mqtt_connected:
            self.mqtt.disconnect()
            self.mqtt.loop_stop()
        sys.exit(0)


def env_bool(env: str, false: str = "") -> str:
    """Return env variable or empty string if the variable contains 'false' or is empty."""
    return os.getenv(env.upper(), "").lower().replace("false", "") or false


if __name__ == "__main__" and len(sys.argv) > 2:
    rtsp = RtspEvent()
    if rtsp.type == "READY":
        rtsp.pub_start()
    elif rtsp.type == "READ":
        rtsp.read_start()
